File: web/src/api/preview.ts
import {
  APIEndpoints,
  BaseAPIPayload,
  GetPreviewStatsAPIPayload,
  PreviewAPIs,
} from '../types/api/api';
import { LogSeverity } from '../types/manager/manager';
import { getLfManager } from '../utils/common';
/// @ts-ignore
import { api } from '/scripts/api.js';

export const PREVIEW_API: PreviewAPIs = {
  //#region clearCache
  clearCache: async () => {
    const lfManager = getLfManager();

    const payload: BaseAPIPayload = {
      message: '',
      status: LogSeverity.Info,
    };

    try {
      const response: Response = await api.fetchApi(APIEndpoints.ClearPreviewCache, {
        method: 'POST',
      });

      const code = response.status;

      switch (code) {
        case 200:
          const p: BaseAPIPayload = await response.json();
          if (p.status === 'success') {
            payload.message = p.message || 'Preview cache cleared successfully.';
            payload.status = LogSeverity.Success;
          }
          break;
        case 403:
          payload.message = 'Permission denied: Unable to delete preview cache.';
          payload.status = LogSeverity.Error;
          break;
        default:
          {
            const errorText = await response.text().catch(() => '');
            payload.message = `Unexpected response from the clear-preview-cache API (${code}): ${
              errorText || response.statusText
            }`;
          }
          payload.status = LogSeverity.Error;
          break;
      }
    } catch (error) {
      payload.message = String(error);
      payload.status = LogSeverity.Error;
    }

    lfManager.log(payload.message, { payload }, payload.status);
    return payload;
  },
  //#endregion

  //#region getStats
  getStats: async () => {
    const lfManager = getLfManager();

    const payload: GetPreviewStatsAPIPayload = {
      data: {
        total_size_bytes: 0,
        file_count: 0,
        path: '',
      },
      message: '',
      status: LogSeverity.Info,
    };

    try {
      const response: Response = await api.fetchApi(APIEndpoints.GetPreviewStats, {
        method: 'POST',
      });

      const code = response.status;

      switch (code) {
        case 200:
          const p: GetPreviewStatsAPIPayload = await response.json();
          if (p.status === 'success') {
            payload.data = p.data;
            payload.message = p.message || 'Preview stats retrieved successfully.';
            payload.status = LogSeverity.Success;
          }
          break;
        default:
          {
            const errorText = await response.text().catch(() => '');
            payload.message = `Unexpected response from the get-preview-stats API (${code}): ${
              errorText || response.statusText
            }`;
          }
          payload.status = LogSeverity.Error;
          break;
      }
    } catch (error) {
      payload.message = String(error);
      payload.status = LogSeverity.Error;
    }

    lfManager.log(payload.message, { payload }, payload.status);
    return payload;
  },
  //#endregion
};

File: modules/api/preview.py
import os
import shutil

from aiohttp import web
from server import PromptServer

from ..utils.constants import API_ROUTE_PREFIX, EXTERNAL_PREVIEW_SUBDIR
from ..utils.helpers.comfy import get_comfy_dir

# region get-preview-stats
@PromptServer.instance.routes.post(f"{API_ROUTE_PREFIX}/get-preview-stats")
async def get_preview_stats(request):
    """
    Calculate disk usage statistics for the external preview cache.

    Returns:
        JSON response with total size in bytes and file count, or error.
    """
    try:
        base_input = get_comfy_dir("input")
        preview_dir = os.path.join(base_input, EXTERNAL_PREVIEW_SUBDIR)

        total_size = 0
        file_count = 0

        if os.path.exists(preview_dir) and os.path.isdir(preview_dir):
            for root, _, files in os.walk(preview_dir):
                for file in files:
                    file_path = os.path.join(root, file)
                    try:
                        total_size += os.path.getsize(file_path)
                        file_count += 1
                    except OSError:
                        # Skip files we can't access
                        pass

        return web.json_response(
            {
                "status": "success",
                "message": "Preview statistics retrieved successfully.",
                "data": {
                    "total_size_bytes": total_size,
                    "file_count": file_count,
                    "path": preview_dir,
                },
            },
            status=200,
        )

    except Exception as e:
        return web.Response(status=500, text=f"Error: {str(e)}")
# endregion

# region clear-preview-cache
@PromptServer.instance.routes.post(f"{API_ROUTE_PREFIX}/clear-preview-cache")
async def clear_preview_cache(request):
    """
    Delete the entire external preview cache directory.

    Returns:
        JSON response confirming deletion or error.
    """
    try:
        base_input = get_comfy_dir("input")
        preview_dir = os.path.join(base_input, EXTERNAL_PREVIEW_SUBDIR)

        if not os.path.exists(preview_dir):
            return web.json_response(
                {
                    "status": "success",
                    "message": "Preview cache directory does not exist. Nothing to clear.",
                },
                status=200,
            )

        # Remove the entire directory tree
        shutil.rmtree(preview_dir)

        return web.json_response(
            {
                "status": "success",
                "message": f"Preview cache cleared successfully from {preview_dir}",
            },
            status=200,
        )

    except PermissionError:
        return web.Response(
            status=403,
            text="Permission denied: Unable to delete preview cache directory.",
        )
    except Exception as e:
        return web.Response(status=500, text=f"Error: {str(e)}")
# endregion
